### PR TITLE
Workaround for broken Eaton (and others?) string language ID descriptor

### LINKED ISSUE #1925: Eaton 9E2000I unknown device and ups model
Using Unraid apcd working fine , but tryed plugin -NUT - Network UPS Tools by SimonFair .
and having some valus unknow like model and device and missing serial number ...

image

### PR DESCRIPTION
**Pull request by tormodvolden**
@masterwishx has tested this with his Eaton device, both on libusb0 and libusb1. Thanks!

The first two commits revert something that was an unsuccessful attempt at fixing something that turned out to be a buggy device. A more effective workaround is suggested in commits 4-5, which also tidies up and unifies things. Commit 3 reverts "retrials" for easier refactoring in 4-5 and similar functionality is added back in a more unified way in commit 4. Commits 6-7-8 are just tidying up things for future work, not strictly needed.

Fixes #1925 
References #414

TODOs fixed in rework:
- [X] retries should be in nut_get_string and not in API wrapper
- [X] usb1 *_simple should just be kept as simple libusb0/1 API wrapper (if needed at all)

Ideas for future work in other PRs:
- all drivers should use "nut_get_string" instead of *_simple and *_ascii
- langid should be cached somewhere (in USBDevice_t or a new USBHandle_t ?)

### CODE REVIEW

**Comment by masterwishx** [comment]
> @masterwishx Can you please test this with your Eaton device?

Wow , Thanks a lot , sure i will test it ...

**Comment by masterwishx** [comment]
I will try to compile without unraid nut plugin as author of plugin don't have time for it...

**Comment by masterwishx** [comment]
Can I test now?

**Comment by tormodvolden** [comment]
Yes. I had only done minor/optical refinements.

**Comment by masterwishx** [comment]
Also maybe `Eaton 9E 2000i` should be instead of `Eaton 9 2000` ... and `PbAc` instead of `PbA` ...
 
 it seems like the last character is missing in values, also in serial number

**Comment by tormodvolden** [comment]
Thanks for your testing. I probably made an off-by-one error in the string decoding. I'll fix that later.

**Comment by tormodvolden** [comment]
The missing last character should be fixed now.

**Comment by masterwishx** [comment]
@jimklimov when you will have time, please check it, the issue is solved :)

**Comment by jimklimov** [comment]
Sounds great, thanks a lot! I'll take a closer look later though.

Interesting how `battery.voltage(.nominal)` was missing and came back after the off-by-one error fix. Was it supposed to be impacted? Not a string, right?..

**Comment by jimklimov** [comment]
@tormodvolden : do you expect this fix is generic enough to drop the need for older `*langid*` options?

**Comment by masterwishx** [comment]
> Sounds great, thanks a lot! I'll take a closer look later though.
> 
> Interesting how `battery.voltage(.nominal)` was missing and came back after the off-by-one error fix. Was it supposed to be impacted? Not a string, right?..

I'm not sure but maybe `battery.voltage(.nominal)` related to your fix for Eaton 9E names, so when name was not fully then it was not catched?

**Comment by masterwishx** [comment]
@tormodvolden do you need to finish also `TODO` befor merging to main?

**Comment by tormodvolden** [comment]
@masterwishx I have reworked it to address the most important TODO items. It should now work equally well with libsb-1.0 and libusb0 so I would appreciate if you can test both.

@jimklimov The USB code in nut seems complicated by the burden of supporting both libusb0 and libusb1. What are the reasons for supporting libusb0?

There seems to be drivers using the libusb0.c and libusb1.c through the "API" and there seems to be some drivers that do their own stuff. And an outlier tools/nut-scanner/scan_usb.c which does its own dlopen() fun and supports both. Then there are wrappers trying to unify the APIs but not being consistently used. Unifying all this and tidying up would probably help in the long run but could be a lot of work.

I would suggest as a next step to get rid of all direct use of usb_get_string_simple() and libusb_get_string_descriptor_ascii() and instead use the "API" functions or the new common nut_usb_get_string().

**Comment by tormodvolden** [comment]
> @tormodvolden : do you expect this fix is generic enough to drop the need for older `*langid*` options?

The fix in this PR doesn't honour the "langid_fix" option, but will always try to auto-detect by reading out the language ID. Having a way to override the fallback value of en_US could possibly be useful. This option should be a generic one, not limited to a single driver. I don't know if anyone used this option to set anything else than en_US (0x0409). If it only was used for setting 0x0409 and the affected devices (only seen in nutdrv_qx and blazer_usb) won't get hosed up by a langid request (but only return an invalid descriptor) then this option can be dropped now.

I am not sure the "noscanlangid" option adds much value, unless some broken device get hosed up by such requests. Other code is already "scanning" it (reading out the language ID), like is being done here in nut_usb_get_string() as well. The option might be just leftovers from previous work. Please check with those who use it (only seen in nutdrv_qx).

**Comment by masterwishx** [comment]
> I have reworked it to address the most important TODO items. It should now work equally well with libsb-1.0 and libusb0 so I would appreciate if you can test both.

Sure I can test it, but don't know how to test both?

**Comment by tormodvolden** [comment]
> Sure I can test it, but don't know how to test both?

After a fresh checkout or running "make distclean":
```
mkdir build0 && cd build0 && ../configure --with-usb=libusb-0.1 && make
mkdir build1 && cd build1 && ../configure --with-usb=libusb-1.0 && make
```

**Comment by jimklimov** [comment]
As for testing, if the libusb-0.1-dev or similar is available, you can `configure --with-libusb=0.1 && make`

**Comment by tormodvolden** [comment]
Let me know if there is any platform where nut builds but there is no libusb-1.0.

**Comment by masterwishx** [comment]
> > Sure I can test it, but don't know how to test both?
> 
> After a fresh checkout or running "make distclean":
> 
> ```
> mkdir build0 && cd build0 && ../configure --with-usb=libusb-0.1 && make
> mkdir build1 && cd build1 && ../configure --with-usb=libusb-1.0 && make
> ```

What I compiled in Slackware VM last time for test was nut.SlackBuild from Unraid NUT plugin modified by me. So it was libusb 1.0.26 that my Unraid version has. 

So I can test latest PR tomorrow same way, but for libusb 0.1 I can add `libusb-0.1` to .. /configure but will it work?

**Comment by jimklimov** [comment]
> Let me know if there is any platform where nut builds but there is no libusb-1.0.

The oldest occasionally tested that I know of, because there still is practical interest, is Solaris 8. Gotta check if that involves USB, and which, though.

**Comment by masterwishx** [comment]
> I have reworked it to address the most important TODO items. It should now work equally well with libsb-1.0 and libusb0 so I would appreciate if you can test both.

Compiled with `libusb 1.0` , all working fine as before , all values available . 

will check with `libusb 0.1` as Unraid have `libusb-compat 0.1.7`

**Comment by masterwishx** [comment]
Compiled also with libusb-0.1 also all working fine ,all values available:  

```
battery.capacity: 7.00
battery.charge: 100
battery.charge.low: 20
battery.charge.restart: 0
battery.protection: yes
battery.runtime: 3246
battery.runtime.low: 180
battery.type: PbAc
battery.voltage: 83.0
battery.voltage.nominal: 72
device.mfr: EATON
device.model: 9E2000i
device.serial:  xxxxxxxxxxxx                   #REDACTED by me # !!! serial is shown now !!!
device.type: ups
driver.debug: 0
driver.flag.allow_killpower: 0
driver.name: usbhid-ups
driver.parameter.pollfreq: 30
driver.parameter.pollinterval: 2
driver.parameter.port: auto
driver.parameter.synchronous: auto
driver.state: quiet
driver.version: 2.8.2.1
driver.version.data: MGE HID 1.48
driver.version.internal: 0.56
driver.version.usb: libusb-0.1 (or compat)
input.bypass.frequency: 49.9
input.bypass.voltage: 231.0
input.frequency: 49.9
input.frequency.nominal: 50
input.transfer.high: 300
input.transfer.low: 100
input.voltage: 231.0
input.voltage.nominal: 231
outlet.1.status: on
outlet.desc: Main Outlet
outlet.id: 0
outlet.switchable: no
output.current: 1.80
output.frequency: 49.9
output.frequency.nominal: 50
output.voltage: 229.0
output.voltage.nominal: 230
ups.beeper.status: enabled
ups.delay.shutdown: 20
ups.delay.start: 30
ups.firmware: 00.02.000
ups.load: 21
ups.load.high: 105
ups.mfr: EATON
ups.model: 9E2000i
ups.power: 424
ups.power.nominal: 2000
ups.productid: ffff
ups.realpower: 303
ups.realpower.nominal: 1600
ups.serial:  xxxxxxxxxxxx                   #REDACTED by me # !!! serial is shown now !!!
ups.shutdown: enabled
ups.start.auto: yes
ups.start.battery: yes
ups.start.reboot: yes
ups.status: OL
ups.temperature: 24.9
ups.test.interval: 604800
ups.test.result: Done and passed
ups.timer.shutdown: -1
ups.timer.start: -1
ups.type: online
ups.vendorid: 0463
```

**Comment by jimklimov** [comment]
Looks great, thank you very much!

Note that (IIRC) libusb-compat is a variant of libusb-1.0 internals which expose libusb-0.1 API as a stop-gap for quickly porting stuff like NUT to use the modern library code.

Still, as far as NUT is concerned, its own libusb0.c/h layer apparently worked well with these changes.

**Comment by desertwitch** [comment]
Many thanks for the hard work, will be happy to integrate this into the Unraid plugin also as soon as it's merged.
Apologies for not being of any help while on holidays and my perhaps a bit snappy response in the related issue...

**Comment by jimklimov** [comment]
Sure, will merge as soon as I can review the solution :)

File: drivers/usb-common.c
Comment (masterwishx) [inline drivers/usb-common.c:422]: it seems we need to init value to pass CI
`int ret = -1;`

Comment (tormodvolden) [reply drivers/usb-common.c:422]: Thanks! I am not able to see the details of the CI without granting a third-party company write access to my repositories! No way! It is ridiculous because all details of the CI are public. I also wonder why this CI of doom runs 4 hours when it already finds a blocker early on.

As you can see from the code, there is no possibility of "ret" being used uninitialized, so there is absolutely no need to initialize it explicitly. But I am not going to fix the CI, so I am applying your hotfix, thanks again.

Comment (jimklimov) [reply drivers/usb-common.c]: Interesting, thanks for the report/complaint.

The CI is a rather standard Jenkins with a GitHub auth plugin, so if it asks for too much to let one in - perhaps there's a way for smarter setup to require less.

As for the time - it runs different build scenarios on roughly a dozen platforms with their different built-in nittiness preferences even for same-named same-version compilers. Stopping on a first error makes little sense, as the builders made 15-20 tech-years apart choke on different things. Sometimes opposite ones, like wanting `switch/case` to cover all known values of an enum, to always have a `default` label, and disliking the label when used with an enum where all values were handled. Alas, that led to a significant amount of pragmas, albeit needed in a handful of code patterns. But, NUT still builds and works everywhere it did earlier...

Comment (jimklimov) [reply drivers/usb-common.c]: Also, in Github links to broken tests, it should have URLs to that specific failed case's logs on CI. Could you please check - do those artifacts now also require a login with "too many permissions required", or work anonymously too?

Comment (tormodvolden) [reply drivers/usb-common.c]: I am not able to find back to the failed jobs any longer. But it looked like all links to the external CI required a login. There was an "aggregated" summary on GitHub saying there was a "conditional-uninitialized" warning but not specifying where.

**Comment by jimklimov** [comment]
I'll prefer some quality time with a PC to actually review this; but at least the different CI platforms favour this change now :)

I guess the static analyzers in compilers were worried that a different setting of `retries` variable or compiler optimization could end up in skipping the loop and returning stack garbage.

**Review by jimklimov — CHANGES_REQUESTED** [review]
LGTM and quite solid.

I've posted some specific questions and points I'd rather see double-checked or changed (might do myself when I get to a computer). Also, an entry in the `NEWS.adoc` file would be good (and in `UPGRADING.adoc`, especially if some of those older `*langid*` settings are useless now), and in `acknowledgements.txt` especially if this were a corporate-backed contribution.

File: drivers/libusb1.c
Comment (tormodvolden) [reply drivers/libusb1.c:978]: Hi, thanks for your review! Do you review commit-per-commit or only the final diff? It might be easier to follow if going through the commits.

I should point out here that the comment changes in "Fix up comments about expected types" update the comments to reflect the status quo (same before or after my changes, I didn't change anything in this regard). So this commit is technically independent of the other commits. It just makes the following commit easier to digest. Having these corrected comments in front of us, it is easier to see why the range-checks were flawed (and they are not gone, only reduced to what is needed, and at the same time removing the need for all the ugly pragmas).

In libusb1.c, StrIdx is uint8_t, so some code cannot pass 256 to it.

In libusb0.c, StrIdx is int, so for positive values, it cannot be larger than MAX_INT anyway. Maybe the original reason for some compiler warnings that were hidden with pragmas. However since it was cast to uintmax_t, an original negative value could indeed become higher than MAX_INT. But in that case the first < 0 would have caught that, so it is superfluous anyway. So the compiler would still be right to complain. Finally this all is useless anyway, because the highest usable value is 255.

I believe the original intention of the casting and comparing was to make something that would work for libusb0 and libusb1 with identical code. And not thinking about the practical limits of the value.

File: drivers/libusb0.c
Comment (tormodvolden) [reply drivers/libusb0.c:404]: It is technically redundant, because the general logging in nut_usb_get_string_descriptor() tells which descriptor was failing. But I agree it is easier to read something about "Manufacturer" instead of looking looking up what string index "1" was about. And I suppose they should have different log levels, with the one in nut_usb_get_string_descriptor() being only debug (since retries may have solved it) whereas the final failure can be critical for the device identification.

I am not familiar with the debug levels in use (or which direction they go). Can you please suggest suitable levels?

File: drivers/libusb0.c
Comment (tormodvolden) [reply drivers/libusb0.c:478]: MAX_RETRY was introduced in the reverted commit and made sense there because it was used in many places. Now it is only used in one place, and I thought "retries = 3" is as readable as packing it into a macro that is defined somewhere else in the code only to be used once. In any case I think it should be commented that 3 is just an arbitrary number and guesswork, not based on experience or logic.

Technically "3" is the number of tries, not retries. Would you like me to reintroduce MAX_RETRY and make it loop 1 + MAX_RETRY?

Finally, are we sure the retries for string descriptors are actually needed? #414 seems unclear on this, I only understood that the HID descriptors needed retries, not sure about any string descriptors. In general we shouldn't add extra code "just in case, cannot hurt" unless it is has been proven to help at least in one real case.

File: drivers/usb-common.c
Comment (tormodvolden) [reply drivers/usb-common.c:468]: Yes. Unless you know about a device that:
- returns a valid langid &&
- the valid reported langid is not en_US &&  #already here it is gonna be difficult to find examples
- doesn't work with the reported langid &&  #would be an "interesting" bug, would probably fail on most OS'es
- works with en_US

Comment (jimklimov) [reply drivers/usb-common.c:468]: Ok, sounds fair :)

Comment (tormodvolden) [reply drivers/usb-common.c:468]: "In general we shouldn't add extra code "just in case, cannot hurt" unless it is has been proven to help at least in one real case." applies here as well :)

File: drivers/usb-common.c
Comment (tormodvolden) [reply drivers/usb-common.c:476]: Yes, always UTF-16LE. The standard was a bit unclear from the start, just stating "Unicode", but in practice all devices use UTF-16LE. There is a "USB Engineering Change Notice" (UnicodeECN.pdf) from 2005 that specifies UTF-16LE and mentions "There are no known cases of devices using an encoding other than UTF-16LE".

"In general we shouldn't add extra code "just in case, cannot hurt" unless it is has been proven to help at least in one real case." again (I'll shut up now).

File: drivers/libusb1.c
Comment (jimklimov) [reply drivers/libusb1.c:978]: For this review, I went with the final diff as attached to the PR. Reviewing individual in-flight commits proved problematic earlier, as force-pushes detached them from PR and project history.

As for the int type analysis - thanks, I concur. As for pragmas - good riddance :D

File: drivers/libusb0.c
Comment (jimklimov) [reply drivers/libusb0.c:478]: Ok, makes sense. For "being sure" - I can only suggest scrolling back in `git blame` (or GH UI equivalent) likely on `libusb0.c` and check what comments might appear in commit or PR that introduced those?..

File: drivers/usb-common.c
Comment (tormodvolden) [reply drivers/usb-common.c:476]: Note that we do a "simple" decoding like in libusb usb_get_string_simple() and libusb_get_string_descriptor_ascii() (the latter is technically not correctly named because it returns 8-bit and not ASCII) and not a full UTF-16LE decoding. If a full decoding would be needed I suppose it would be better to use some iconv-like library. libusb never heard about any issues with the simple decoding AFAIK.

File: drivers/libusb0.c
Comment (jimklimov) [reply drivers/libusb0.c:404]: The lower debug nunbers are more visible, but so far there are few formal criteria about what is exposed how. I guess the lines in `nut_usb_get_string_descriptor()` can be emitted with 2 or 3, so a hiccup getting those values (an improbable event per se?) can be seen with little noise from other events that would require a higher verbosity to troubleshoot.

File: drivers/usb-common.c
Comment (tormodvolden) [reply drivers/usb-common.c:463]: Yes, I think ideally the fallback value should be configurable, in case some (broken) device shows up and requires another fallback value. But this can safely wait until such a device is proven to exist maybe? I think there will be (more) non-en_US devices in the future, but at the same time their manufacturers will probably make sure their langid's are correct in this case.

BTW, with regards to possibly generalizing e.g. langid_fix. I don't know how nut works with multiple configs and multiple devices. Will each device have its own process, so globals are fine, and will get_var only return the values relevant for this instance?

And no, libusb has no langid mappings or helpers. It doesn't touch langid other than in the _simple/_ascii helpers (apart from possibly some assumptions in the little used Windows HID backend).

File: drivers/libusb0.c
Comment (tormodvolden) [reply drivers/libusb0.c:478]: Oh, in the retries code (if we want to keep it :) ) I added a delay, that was not there before. This is also guesswork "can probably help". But I meant to use 50 ms, not us.

Some arguments for not doing retries are that USB does low-level retries already, and that e.g. any issues with concurrent access from other code or OS should be dealt with properly instead of papering over with retries.

**Comment by tormodvolden** [comment]
BTW, why did the CI kick off again (on the same identical git hash)? Because of the added labels or milestone?

File: drivers/libusb1.c
Comment (tormodvolden) [reply drivers/libusb1.c:978]: Ah, you're reviewing per e-mail. In the web interface it is easy to go through each commit. Does it mean you haven't read the commit messages at all?

File: drivers/libusb0.c
Comment (tormodvolden) [reply drivers/libusb0.c:478]: I did try to look up the reason for the retries, commit 0fa5687a took me to #414 as mentioned in my commit message.

BTW, although you are attributed as author, with kudos to Sam Varshavchik, it was actually @aquette that came up with the code in https://github.com/networkupstools/nut/issues/414#issuecomment-292143944.

Looking at log1.txt and log2.txt from https://github.com/networkupstools/nut/issues/414#issuecomment-291639104 it actually looks like there is a spurious Product string retrieval failure, that was not seen again after applying @aquette's patch. Sure that patch includes another fix as well, but I agree that looks unrelated. Sadly we don't see "retrying..." in the logs, so we cannot for sure know if the patch actually helped.

So I'd say we keep the retries, comment that this is introduced for "Tripp Lite", make sure we log any retries for future evaluation after it has been tested with "Tripp Lite" products for a while.

File: drivers/libusb0.c
Comment (tormodvolden) [reply drivers/libusb0.c:404]: As mentioned in the other thread, if we are not sure if these retries will ever happen (?) then maybe we should make sure we'll always see them in logs, for collecting statistics on it.

File: drivers/libusb0.c
Comment (tormodvolden) [reply drivers/libusb0.c:478]: I see some other code is using MAX_TRIES and I think that is better than 1 + MAX_RETRY, unless you prefer something more specific like MAX_STRING_DESC_TRIES.

File: drivers/libusb0.c
Comment (jimklimov) [reply drivers/libusb0.c:404]: Does make sense too.

File: drivers/libusb0.c
Comment (jimklimov) [reply drivers/libusb0.c:478]: I wonder if those devices had a sort of throttling when overwhelmed with queries, so sleeping helped? A comment makes sense, and a domain-specific macro name too.

File: drivers/usb-common.c
Comment (jimklimov) [reply drivers/usb-common.c:476]: Thanks for clarifications!

File: drivers/libusb1.c
Comment (jimklimov) [reply drivers/libusb1.c:978]: I did read the messages... now :D (thanks for diligently filling those out). But alas not on the first pass by phone from the road.

**Comment by jimklimov** [comment]
> BTW, why did the CI kick off again (on the same identical git hash)? Because of the added labels or milestone?

Changed target branch (another PR merged).

**Comment by tormodvolden** [comment]
> > BTW, why did the CI kick off again (on the same identical git hash)? Because of the added labels or milestone?
> 
> Changed target branch state/content (another PR merged).

But the CI job was rerun on the unchanged PR code, so useless waste of CI hours...

File: drivers/libusb0.c
Comment (tormodvolden) [reply drivers/libusb0.c:478]: There was no sleeping in the original retries, so it is apparently not needed. I just think it can help (and cannot hurt(TM)), for instance if used with faster computers than originally tested on.

**Comment by tormodvolden** [comment]
I pushed some fixups (to keep your CI busy ;) ), I'll squash these with their respective base commits and rebase to current master once we agree on all details.

**Comment by jimklimov** [comment]
> > > BTW, why did the CI kick off again (on the same identical git hash)? Because of the added labels or milestone?
> > 
> > 
> > Changed target branch state/content (another PR merged).
> 
> But the CI job was rerun on the unchanged PR code, so useless waste of CI hours...

Well, sometimes people (or myself with different hats on) poke same areas of code, recipes, docs etc. so there may be merge conflicts or logical mismatches that pop up in tests.

The merge job builds use an ephemeral branch for that (what-if this PR got merged into target now?)

**Comment by tormodvolden** [comment]
> The merge job builds use an ephemeral branch for that (what-if this PR got merged into target now?)

Ah OK. But I though I would see another hash then, corresponding to the ephemeral merge commit.

**Comment by masterwishx** [comment]
@tormodvolden can you see error in `All issues details`?

**Comment by jimklimov** [comment]
An odd hiccup - agent ran out of temp space, likely. Says some sources were missing.

**Comment by masterwishx** [comment]
i can see next errors : `Trying to fetch newest commits from unstashed archive provided by the build` 

check it here: https://github.com/networkupstools/nut/pull/2604/checks?check_run_id=29553874289

image

**Comment by jimklimov** [comment]
Just wanted to say something that was on my mind for a week: initially I think @masterwishx seemed uncertain about how to get NUT custom-built to research for this issue, what the tools are and how to get them, and now is quite comfortable trying out the builds and pointig to issues. And brought @tormodvolden onboard, too.

Community-wise, I think those are all even greater achievements than nailing a particular technical issue.

**Comment by jimklimov** [comment]
Hm, maybe an older git implem there? Others did not complain for same activity (using a stashed initial checkout as a reference repo).

"unadvertised object" huh...

**Comment by masterwishx** [comment]
> Just wanted to say something that was on my mind for a week: initially I think @masterwishx seemed uncertain about how to get NUT custom-built to research for this issue, what the tools are and where to find them, and now is quite comfortable trying out the builds and pointing to issues. And brought @tormodvolden onboard, too.
> 
> Community-wise, I think those are all even greater achievements than nailing a particular technical issue.

Just wanted to point to @tormodvolden where to see the issue without Auth to CI Jenkins as he told he can't see where the error.

**Review by jimklimov — APPROVED** [review]

**Comment by jimklimov** [comment]
Bumped the PR with a NEWS entry and resync with current master.

@tormodvolden : if you wish to squash some of the commits and rebase - feel free, just re-fetch to have this NEWS change attached. Or I can just merge as soon as CI does not bork on anything relevant. Law of big numbers: something random does hiccup eventually.

**Comment by jimklimov** [comment]
````
make[1]: *** [Makefile:1644: install-win-bundle-thirdparty] Error 1

make: *** [Makefile:1638: install-win-bundle] Error 2
````

Not sure what Appveyor disliked here, build/test were ok. So this fault is unrelated to this PR.

FWIW, it listed some DLLs that were not in mingw (FOSS) locations but that's normal (e.g. OS own libs). Not sure if the list is same as before, or is why the step failed. Maybe `libexec/sockdebug.exe` troubleshooting helper being not covered by DLL accompaniment?..

File: drivers/usb-common.c
Comment (aquette) [reply drivers/usb-common.c:468]: Happy to see that things are good.
Congrats for the team work.
And thanks for the credits @tormodvolden

**Comment by masterwishx** [comment]
> ```
> make[1]: *** [Makefile:1644: install-win-bundle-thirdparty] Error 1
> 
> make: *** [Makefile:1638: install-win-bundle] Error 2
> ```
> 
> Not sure what Appveyor disliked here, build/test were ok. So this fault is unrelated to this PR.
> 
> FWIW, it listed some DLLs that were not in mingw (FOSS) locations but that's normal (e.g. OS own libs). Not sure if the list is same as before, or is why the step failed. Maybe `libexec/sockdebug.exe` troubleshooting helper being not covered by DLL accompaniment?..

Also I saw it was fine in test befor...

**Comment by masterwishx** [comment]
Finally merge is possible

**Pull request merged**
